File: mobile/android/app/src/test/java/com/radiostream/player/PlaylistPlayerTest.java
 package com.radiostream.player;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.radiostream.networking.metadata.MetadataBackendGetter;
import com.radiostream.ui.PlayerNotification;
import com.radiostream.util.SetTimeout;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import static com.radiostream.player.Utils.resolvedPromise;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(PowerMockRunner.class)
@PrepareForTest({Arguments.class, android.util.Log.class})
public class PlaylistPlayerTest {

    @Mock
    Song mockFirstSong;

    @Mock
    Song mockSecondSong;

    @Mock
    Playlist mockPlaylist;

    @Mock
    PlaylistFactory mockPlaylistFactory;

    @Mock
    MetadataBackendGetter mockMetadataBackendGetter;

    @Mock
    SetTimeout mockSetTimeout;

    @Mock
    StatusProvider mockStatusProvider;

    @Mock
    PlayerNotification mockPlayerNotifcation;


    @Before
    public void setUp() throws Exception {
        Utils.initTestLogging();
        Utils.mockAndroidStatics();

        final WritableMap dummyFirstSongBridge = Arguments.createMap();
        dummyFirstSongBridge.putString("title", "mockFirstSong");
        when(mockFirstSong.toBridgeObject()).thenReturn(dummyFirstSongBridge);

        final WritableMap dummySecondSongBridge = Arguments.createMap();
        dummyFirstSongBridge.putString("title", "mockSecondSong");
        when(mockSecondSong.toBridgeObject()).thenReturn(dummySecondSongBridge);

        when(mockFirstSong.preload()).thenReturn(resolvedPromise(mockFirstSong));
        when(mockSecondSong.preload()).thenReturn(resolvedPromise(mockSecondSong));
        when(mockFirstSong.waitForMarkedAsPlayed()).thenReturn(resolvedPromise(true));
        when(mockSecondSong.waitForMarkedAsPlayed()).thenReturn(resolvedPromise(true));

        when(mockPlaylist.peekCurrentSong())
            .thenReturn(resolvedPromise(mockFirstSong));

        when(mockPlaylist.peekNextSong())
            .thenReturn(resolvedPromise(mockSecondSong));

        when(mockPlaylist.isCurrentSong(mockFirstSong)).thenReturn(true);

        final WritableMap dummyPlaylistBridge = Arguments.createMap();
        dummyFirstSongBridge.putString("name", "x");
        when(mockPlaylist.toBridgeObject()).thenReturn(dummyPlaylistBridge);
    }

    @Test
    public void play_playSongIfSongAvailable() throws Exception {
        PlaylistPlayer playlistPlayer = new PlaylistPlayer(mockPlaylist, mockSetTimeout, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation);
        playlistPlayer.play();

        verify(mockPlaylist, atLeastOnce()).peekCurrentSong();
        verify(mockFirstSong, atLeastOnce()).play();

        playlistPlayer.play();
        verify(mockFirstSong, Mockito.times(1)).preload();

    }

    @Test
    public void playNext_playingSecondSong() throws Exception {
        PlaylistPlayer playlistPlayer = new PlaylistPlayer(mockPlaylist, mockSetTimeout, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation);
        playlistPlayer.playNext();

        verify(mockPlaylist, times(1)).nextSong();
        verify(mockFirstSong, times(1)).play();
    }

    @Test
    public void playNext_playFirstSong() throws Exception {
        PlaylistPlayer playlistPlayer = new PlaylistPlayer(mockPlaylist, mockSetTimeout, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation);
        playlistPlayer.playNext();

        verify(mockFirstSong, times(1)).play();
    }

    @Test(expected=IllegalStateException.class)
    public void pause_throwsExceptionIfNoSong() throws Exception {
        PlaylistPlayer playlistPlayer = new PlaylistPlayer(mockPlaylist, mockSetTimeout, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation);
        playlistPlayer.pause();
    }

    @Test
    public void close_closesPlaylist() throws Exception {
        PlaylistPlayer playlistPlayer = new PlaylistPlayer(mockPlaylist, mockSetTimeout, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation);
        playlistPlayer.close();
    }

    @Test
    public void close_closesSongIfExists() throws Exception {
        PlaylistPlayer playlistPlayer = new PlaylistPlayer(mockPlaylist, mockSetTimeout, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation);
        playlistPlayer.play();
        playlistPlayer.close();

        verify(mockFirstSong, times(1)).close();
    }

    @Test
    public void playNext_retriesOnFailure() throws Exception {
        when(mockFirstSong.preload())
            .thenReturn(Utils.<Song>rejectedPromise(new Exception()))
            .thenReturn(Utils.resolvedPromise(mockFirstSong));

        when(mockSetTimeout.run(anyInt())).thenReturn(resolvedPromise((Void)null));

        PlaylistPlayer playlistPlayer = new PlaylistPlayer(mockPlaylist, mockSetTimeout, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation);
        playlistPlayer.play();

        // song will be loaded if preloading the first one failed
        verify(mockFirstSong, times(1)).play();
    }
}

File: app/android/app/src/test/java/com/radiostream/player/PlaylistPlayerTest.kt
package com.radiostream.player

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.times
import com.nhaarman.mockito_kotlin.whenever
import com.radiostream.networking.metadata.MetadataBackendGetter
import com.radiostream.ui.PlayerNotification
import kotlinx.coroutines.experimental.runBlocking
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.mockito.Mockito.atLeastOnce
import org.mockito.Mockito.verify


class PlaylistPlayerTest {

    @Before
    @Throws(Exception::class)
    fun setUp() {
        Utils.initTestLogging()
    }

    suspend fun returnMockSongData(song: Song) {

        val dummyFirstSongBridge = ArgumentsWrapperTest().createMap()
        dummyFirstSongBridge.putString("title", "mockFirstSong")
        whenever(song.toBridgeObject()).thenReturn(dummyFirstSongBridge)

    }

    @Test
    fun play_playSongIfSongAvailable() = runBlocking {

        val mockPlayerNotifcation = mock<PlayerNotification>()
        val mockMetadataBackendGetter = mock<MetadataBackendGetter>()
        val mockStatusProvider = mock<StatusProvider>()

        val mockFirstSong = mock<Song>()
        returnMockSongData(mockFirstSong)

        val mockPlaylist = mock<Playlist>()
        whenever(mockPlaylist.peekCurrentSong()).thenReturn(mockFirstSong)
        whenever(mockPlaylist.isCurrentSong(mockFirstSong)).thenReturn(true)

        val playlistPlayer = PlaylistPlayer(mockPlaylist, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation, ArgumentsWrapperTest())
        playlistPlayer.play()

        verify<Playlist>(mockPlaylist, atLeastOnce()).peekCurrentSong()
        verify<Song>(mockFirstSong, atLeastOnce()).play()

        playlistPlayer.play()
        verify<Song>(mockFirstSong, Mockito.times(1)).preload()
    }

    @Test
    fun playNext_playingNextSong() = runBlocking {
        val mockPlayerNotifcation = mock<PlayerNotification>()
        val mockMetadataBackendGetter = mock<MetadataBackendGetter>()
        val mockStatusProvider = mock<StatusProvider>()

        val mockFirstSong = mock<Song>()
        returnMockSongData(mockFirstSong)

        val mockPlaylist = mock<Playlist>()
        whenever(mockPlaylist.peekCurrentSong()).thenReturn(mockFirstSong)
        whenever(mockPlaylist.isCurrentSong(mockFirstSong)).thenReturn(true)

        val playlistPlayer = PlaylistPlayer(mockPlaylist, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation, ArgumentsWrapperTest())
        playlistPlayer.playNext()

        verify<Playlist>(mockPlaylist, times(1)).nextSong()
        verify<Song>(mockFirstSong, times(1)).play()
    }


    @Test(expected = IllegalStateException::class)
    fun pause_throwsExceptionIfNoSong() = runBlocking {
        val mockPlayerNotifcation = mock<PlayerNotification>()
        val mockMetadataBackendGetter = mock<MetadataBackendGetter>()
        val mockStatusProvider = mock<StatusProvider>()
        val mockPlaylist = mock<Playlist>()

        val playlistPlayer = PlaylistPlayer(mockPlaylist, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation, ArgumentsWrapperTest())
        playlistPlayer.pause()
    }

    @Test
    fun close_closesPlaylistAndSongIfExists() = runBlocking {
        val mockPlayerNotifcation = mock<PlayerNotification>()
        val mockMetadataBackendGetter = mock<MetadataBackendGetter>()
        val mockStatusProvider = mock<StatusProvider>()

        val mockFirstSong = mock<Song>()
        returnMockSongData(mockFirstSong)

        val mockPlaylist = mock<Playlist>()
        whenever(mockPlaylist.peekCurrentSong()).thenReturn(mockFirstSong)
        whenever(mockPlaylist.isCurrentSong(mockFirstSong)).thenReturn(true)

        val playlistPlayer = PlaylistPlayer(mockPlaylist, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation, ArgumentsWrapperTest())
        playlistPlayer.play()
        playlistPlayer.close()

        verify<Song>(mockFirstSong, times(1)).close()
    }

    @Test
    fun playNext_retriesOnFailure() = runBlocking {
        val mockPlayerNotifcation = mock<PlayerNotification>()
        val mockMetadataBackendGetter = mock<MetadataBackendGetter>()
        val mockStatusProvider = mock<StatusProvider>()

        val mockFirstSong = mock<Song>()
        returnMockSongData(mockFirstSong)
        whenever(mockFirstSong.preload()).thenThrow(RuntimeException()).thenReturn(Unit)

        val mockPlaylist = mock<Playlist>()
        whenever(mockPlaylist.peekCurrentSong()).thenReturn(mockFirstSong)
        whenever(mockPlaylist.isCurrentSong(mockFirstSong)).thenReturn(true)

        val playlistPlayer = PlaylistPlayer(mockPlaylist, mockMetadataBackendGetter, mockStatusProvider, mockPlayerNotifcation, ArgumentsWrapperTest())
        playlistPlayer.play()

        // song will be loaded if preloading the first one failed
        verify<Song>(mockFirstSong, times(1)).play()
    }
}